### ISSUE TITLE
memory leak reading large amounts of string data

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#985]
**Reported by ansley b manke on 31 Dec 2009 19:16 UTC**
Kevin noticed Ferret hogging lots of memory on a script that reads lots of string variables on a long I axis.  When put into LAS, which was starting Ferret with the -memsize 16 switch (smaller than the current Ferret default), the process started using more and more memory.

Here is a long report summarizing what I learned about this. The "unblocking" of string data that's been read from netCDF is trying to use unallocated memory, or something. Possibly what is needed is for Ferret to detect this situation (how?) and report a need to ask the user to allocate more memory. 

A workaround is to make sure to start Ferret with a large enough amount of memory.

1) The -memsize 16 switch should be removed from the LAS default startup of Ferret, as the current Ferret default is 25.6 as of Ferret v6.2. I've asked Roland to do that.

2) I Built Ferret with the switches to use the Valgrind memory-check tool. When the array of strings is over 120000 items long, and when memory is set small enough, the simple script below results in memory leak reports when running valgrind with memory checking.

The data file used in the script is too big to attach but is at
/home/porter/ansley/ans_ferret/users/kobrien/gridded/string_type.nc

```
! onemask.jnl
! Simple script based on createGriddedFilesTopLevel.jnl
! Read a string variable on a long axis.
!
! USE "/home/raid1/data/kobrien/data/osmcExtractForGriddedSummary_20091103-20091109.nc"
! save/file=string_type.nc/clobber type

yes? use string_type.nc
yes? sh dat
     currently SET data sets:
    1> ./string_type.nc  (default)
 name     title       I         J         K         L
 TYPE                 1:123329  ...       ...       ...

yes? let drifter_mask = if (strindex(type, "DRIFTERS") ge 1) then 1
yes? load drifter_mask
```

This command to run Valgrind's memcheck tool, starting Ferret with just 6 mwords of memory  results in reports from valgrind of a leak in tm_unblockify_ferret_strings

  /home/nstout/ansley/valgrind/bin/valgrind --leak-check=yes /home/porter/ansley/valgrind_ferret/FERRET/fer/ferretdods_gui -memsize 6 -gif -server -script onemask.jnl >& valgrind_output_memsize6_onemask.output

The Valgrind output includes:
```

...
== == 13,500,968 bytes in 616,645 blocks are definitely lost in loss record 68 of 70
== ==    at 0x490514E: malloc (vg_replace_malloc.c:195)
== ==    by 0x82A0A8: tm_unblockify_ferret_strings (tm_unblockify_ferret_strings.c:66)
== ==    by 0x829C69: cd_read_sub_ (cd_read_sub.c:183)
== ==    by 0x821AA3: cd_read_ (cd_read.f:1713)
== ==    by 0x81169B: tm_choose_read_ (tm_choose_read.f:217)
== ==    by 0x7F6C81: tm_read_ (tm_read.f:285)
== ==    by 0x691E21: read_tm_ (read_tm.f:271)
== ==    by 0x5F5112: is_read_ (is_read.f:312)
== ==    by 0x5EC455: interp_stack_ (interp_stack.f:429)
== ==    by 0x5EB132: get_prot_cmnd_data_ (get_prot_cmnd_data.f:156)
== ==    by 0x5EAEE1: get_cmnd_data_ (get_cmnd_data.f:80)
== ==    by 0x565369: xeq_load_ (xeq_load.f:98)
== == 

```
For this simple example, starting Ferret with more memory at the start, such as the default memory size, results no reports about leaks in tm_unblockify_ferret.  Likewise if the dimension of the string variable is smaller, say 50000 strings in the input file, this error does not occur.

Now, tm_unblockify_ferret_strings.c is unpacking the string data read from a netCDF string-data read.  Here is the code. Record 66 reported by valgrind is the line that contains the malloc of variable poutchar.
```
/home/porter/ansley/valgrind_ferret/FERRET/fmt/src/tm_unblockify_ferret_strings.c

/*
    The "ferdat" variable is a pointer to an array of string pointers
      where the string pointers are spaced 8 bytes apart
    The buffer pbuff is a long character array. Comy the strings to pbuff
      spacing the strings outstrlen apart.
*/
*/

/* *kob* 10/03 v553 - gcc v3.x needs wchar.h included */
#include <wchar.h>
#include <stdlib.h>

void tm_unblockify_ferret_strings(char **mr_blk1, char *pblock,
				int bufsiz, int filestrlen)
{
  int i, ii, n;
  char *pinchar, *pinstr, *poutchar, **poutstr;

  /* copy all the strings */
  poutstr = mr_blk1;  /* points to each output string in turn */
  pinstr  = pblock;   /* points to each input  string in turn */
  for (i=0; i<bufsiz/filestrlen; i++) {

    /* measure the length of the string */
    n = 0;
    while (n<filestrlen && pinstr[n++;

    pinchar = pinstr;
    pinstr += filestrlen;       /* ready for next input string */

    /* allocate memory for string */
    poutchar = (char *) malloc(sizeof(char) * (n+1)); 
    *poutstr = poutchar;
    poutstr += (8/sizeof(char**));   /* pointer to next output string */
     
    /* copy 1 string */
    n = 0;
    while (n<filestrlen && *pinchar) {
      *poutchar = *pinchar;
      poutchar++;
      pinchar++;
      n++;
    }
    *poutchar = NULL;

  }

   return;
}

```

There is a malloc of poutchar, but no free - but it's doing tricky things with the string and pointer variables and so simply adding a free(poutchar) command does not work. Tried replacing the  
*poutchar = NULL; with a free(poutchar) but this results in Valgrind reporting that:

```
== == Invalid free() / delete / delete[](n]))
== ==    at 0x4905E72: free (vg_replace_malloc.c:325)
== ==    by 0x82A103: tm_unblockify_ferret_strings (tm_unblockify_ferret_strings.c:78)
```

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/1713

**Comment by karlmsmith** [comment]
**Modified by ansley.b.manke on 5 Jan 2011 19:49 UTC**

**Comment by karlmsmith** [comment]
**Comment by karl.smith on 27 Feb 2012 17:48 UTC**
This still appears to be an issue.  (Another script from Kevin using lots of string arrays consuming lots of memory.)  The above code does have a memory leak in that the line ```*poutstr = poutchar``` is overwriting a pointer (if not NULL) to an allocated string that has not been freed.

Elevated the priority to get this back on the table.  Andrew has commented to me that he does use lots of string arrays, so we need to keep him in the loop on this as well.

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 27 Feb 2012 18:52 UTC**
When fixing this bug it would be good to straighten out the documentation at the start of tm_unblockify_ferret_strings.c and tm_blockify_ferret_strings.c.  These routines do not self-explain what their purpose is, and they refer to variables ferdat and pbuf that seem not to exist.

**Comment by karlmsmith** [comment]
**Comment by karl.smith on 19 Mar 2012 20:21 UTC**
tm_unblockify_strings and tm_blockify_strings cleaned up (including comments) in ferret6d branch.  Verified that the "block" string being given to these functions was being freed when no longer needed (soon after these functions were called).

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 21 Mar 2012 22:29 UTC**
In addressing #1466, we removed a call to free_c_string_array in delete_variable.F.  That caused this leak. With that call restored the script that uses lots of string variables and that was hogging so much memory runs correctly and uses just a modest amount of memory.

Because the bug of #1466 was only on one machine we are thinking that some detail of the memory allocation/freeing code was causing a leak on that combination of machine/compiler.  We never saw the behavior elsewhere. Marking this closed.
